Dockerfile

```diff
@@ -1,19 +1,52 @@
-FROM node:12.22.10-alpine3.15@sha256:f150ebf9402f0dd6a9c4cb208ed64884cfa7c8a6ccae3f749a7b12156c25ad88
+FROM node:12.22.10-alpine3.15@sha256:f150ebf9402f0dd6a9c4cb208ed64884cfa7c8a6ccae3f749a7b12156c25ad88 as builder
+### Needed to run pact-mock-service
+COPY sgerrand.rsa.pub /etc/apk/keys/sgerrand.rsa.pub
+RUN ["apk", "--no-cache", "add", "ca-certificates", "python2", "build-base", "bash", "ruby"]
+RUN wget https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.28-r0/glibc-2.28-r0.apk && apk add --no-cache glibc-2.28-r0.apk && rm -f glibc-2.28-r0.apk
+###
+WORKDIR /app
+COPY package.json .
+COPY package-lock.json .
+RUN npm ci --quiet
+
+COPY .snyk .
+COPY Gruntfile.js .
+COPY app/assets ./app/assets
+COPY app/browsered.js ./app/browsered.js
+COPY app/utils ./app/utils
+RUN npm run compile && npm run snyk-protect
+
+
+FROM builder as tester
+WORKDIR /app
+COPY . .
+COPY --from=builder /app/node_modules ./node_modules
+COPY --from=builder /app/govuk_modules ./govuk_modules
+COPY --from=builder /app/public ./public
+RUN npm test
+
+FROM node:12.22.10-alpine3.15@sha256:f150ebf9402f0dd6a9c4cb208ed64884cfa7c8a6ccae3f749a7b12156c25ad88 as final
 
 RUN ["apk", "--no-cache", "upgrade"]
 
 RUN ["apk", "add", "--no-cache", "tini"]
 
-ADD package.json /tmp/package.json
-ADD package-lock.json /tmp/package-lock.json
-RUN cd /tmp && npm ci --production
+WORKDIR /app
+COPY package.json .
+COPY start.js .
+COPY server.js .
+COPY config/ ./config
```

Comment:
Having which folders are required to run the app in the dockerfile feels fairly brittle and tricky to follow (especially if trying to add new things based on current patterns).

Often there is either a single build or configuration with an output folder but this moves it into Docker specific language.

While we generally don't favour copying everything (unpredictable, security, can rely on host machine, etc.) until frontend has a clearer set of production files I think it would be easier to read and less flakey in the future.

A longer term solution to this would be to move everything needed for running the service into a single top level folder (app) and just make sure this is in the final layer.